src/parse.rs

```diff
@@ -99,116 +99,118 @@ impl AST {
-                continue;
-            }
+    for dir in dirs {
+        // Generate module constants
+        let path = dir.path.as_path().to_str().unwrap_or("").to_owned();
 
-            // Generate module identifier
-            let p = std::path::PathBuf::from(path.clone());
-            let mod_path;
-            if p.is_file() {
-                let p: PathBuf = p.into_iter().dropping_back(1).collect();
-                mod_path = p.into_os_string().into_string().unwrap()
-            } else {
-                mod_path = String::from(path.clone());
-            }
-            let module_name = mod_path.clone();
-
-            // Get directory
-            let read_dir = std::fs::read_dir(dir.clone())?;
-            let mut module = ModuleComponent::new(module_name.to_string(), path);
-
-            for entry in read_dir {
-                let entry = entry?;
-                if !entry.path().is_dir() {
-                    let mut file = File::open(entry.path())?;
-                    let (components, lang) = match parse_file(&mut file, &entry.path()) {
-                        Ok(res) => res,
-                        Err(err) => {
-                            eprintln!("Could not read file {:?}: {:#?}", entry.path(), err);
-                            continue;
-                        }
-                    };
-                    language = lang;
-
-                    for component in components.into_iter() {
-                        match component {
-                            ComponentType::ClassOrInterfaceComponent(component) => {
-                                match component.declaration_type {
-                                    ContainerType::Class => {
-                                        module.classes.push(component);
-                                    }
-                                    ContainerType::Interface => {
-                                        module.interfaces.push(component);
-                                    }
-                                    r#type => {
-                                        println!("got other label when it should have been class/ifc: {:#?}", r#type);
-                                    }
+        // Generate module identifier
+        let p = dir.path.clone();
+        let mod_path;
+        if p.is_file() {
+            let p: PathBuf = p.into_iter().dropping_back(1).collect();
+            mod_path = p.into_os_string().into_string().unwrap()
+        } else {
+            mod_path = path.clone();
+        }
+        let module_name = mod_path.clone();
+
+        // Get directory
+        let read_dir = std::fs::read_dir(dir.path)?;
+        let mut module = ModuleComponent::new(module_name.to_string(), path);
+
+        for entry in read_dir {
+            let entry = entry?;
+            if !entry.path().is_dir() {
+                if dir
+                    .files
+                    .iter()
+                    .find(|path_buf| path_buf == &&entry.path())
```

Comment:
More of a Rust question: what's going on with the &&? I know & gets a reference, so is this a double reference? If so, why? I'm just curious.

Reply:
The reason this happens is because `.iter()` returns a `Iter<&PathBuf>` like you'd expect, but then invoking `.find()` also adds another reference layer so `path_buf` ends up being a `&&PathBuf`